pam/go-loader/module.c

```diff
@@ -0,0 +1,142 @@
+#include <stdio.h>
+#include <string.h>
+
+/* When a Go shared library is loaded from C, go starts various goroutine
+ * (as init() at first) and if the loading code is then performing a fork
+ * we end up having an undefined behavior and very likely, deadlocks.
+ *
+ * To avoid this we need to ensure that the module is called after the loading
+ * application has forked, and this for sure has happened when the application
+ * calls the PAM functions that a module exposes.
+ *
+ * As per this, here we re-implement the PAM modules functions and for each
+ * action we load the module if it has not been loaded already, otherwise we
+ * dload it and we redirect each call to the loaded library.
+ */
+
+typedef int (*PamHandler)(pam_handle_t *,
+                          int          flags,
+                          int          argc,
+                          const char **argv);
+
+static void
+on_go_module_removed (pam_handle_t *pamh,
+                      void         *go_module,
+                      int           error_status)
+{
+  void (*go_pam_cleanup) (void);
+  *(void **) (&go_pam_cleanup) = dlsym (go_module, "go_pam_cleanup_module");
+  if (go_pam_cleanup)
+    go_pam_cleanup ();
+
+  dlclose (go_module);
+}
+
+static void *
+load_module (pam_handle_t *pamh,
+             const char   *module_path)
+{
+  void *go_module;
+
+  if (pam_get_data (pamh, "go-module", (const void **) &go_module) == PAM_SUCCESS)
+    return go_module;
+
+  go_module = dlopen (module_path, RTLD_LAZY);
+  if (!go_module)
+    return NULL;
+
+  pam_set_data (pamh, "go-module", go_module, on_go_module_removed);
+
+  void (*init_module) (void);
+  *(void **) (&init_module) = dlsym (go_module, "go_pam_init_module");
+  if (init_module)
+    init_module ();
+
+  return go_module;
+}
+
+static inline int
+call_pam_function (pam_handle_t *pamh,
+                   const char   *function,
+                   int           flags,
+                   int           argc,
+                   const char  **argv)
+{
+  char module_path[PATH_MAX] = {0};
+  const char *sub_module;
+  PamHandler func;
+  void *go_module;
+
+  if (argc < 1)
+    {
+      pam_error (pamh, "%s: no module provided", function);
+      return PAM_MODULE_UNKNOWN;
+    }
+
+  sub_module = argv[0];
+  argc -= 1;
+  argv = (argc == 0) ? NULL : &argv[1];
+
+  snprintf (module_path, PATH_MAX - 1, PAM_MODULES_DIR "/%s", sub_module);
+
+#if DEBUG_MODE
+  if (strchr (sub_module, '/') != NULL)
+    strncpy (module_path, sub_module, PATH_MAX - 1);
+#endif
+
+  go_module = load_module (pamh, module_path);
+  if (!go_module)
+    {
+      pam_error (pamh, "Impossible to load module %s", module_path);
+      return PAM_OPEN_ERR;
+    }
+
+  *(void **) (&func) = dlsym (go_module, function);
+  if (!func)
+    {
+      pam_error (pamh, "Symbol %s not found in %s", function, module_path);
+      return PAM_OPEN_ERR;
+    }
+
+  return func (pamh, flags, argc, argv);
+}
+
+#define DEFINE_PAM_WRAPPER(name) \
```

Comment:
I think we should keep you away from Rust or you will get mad at procedural macros :) (albeit: ok, this one is simple and easy to maintain, thanks for pam to always have the same parameters for all its function!)

Reply:
Cause you're not lucky enough to read my gjs templating ventures in C++ :-D or macro nesting in GLib :)

But yeah... I had it done more complex, then I decided to simplify it. The beauty of C <3.

Reply:
We definitively don’t have the same definition of "lucky" ;)

Reply:
Makes me feel less wonky about GJS :)